feat!: migrate FTS from tantivy to lance-index (#1483)
lancedb/lancedb: docs/src/fts.md

# Full-text search

LanceDB provides support for full-text search via Lance (before via [Tantivy](https://github.com/quickwit-oss/tantivy) (Python only)), allowing you to incorporate keyword-based search (based on BM25) in your retrieval solutions.

> Currently, the Lance full text search is missing some features that are in the Tantivy full text search (custom tokenizer, phrase query, etc.), so in Python, tantivy is still the default way to do full text search, and we are leaving the instructions in place, as we finish adding the rest of this functionality.

Reviewer: This is a very long sentence. Could we split this up?

```suggestion
Currently, the Lance full text search is missing some features that are in the Tantivy full text search. This includes phrase queries, re-ranking, and customizing the tokenizer. Thus, in Python, Tantivy is still the default way to do full text search and many of the instructions below apply just to Tantivy-based indices.
```

## Installation (Only for Tantivy-based FTS)

> 

Reviewer 2: lets add a note that using native FTS does not require install extra dependencies.

Author: Added

## Example

Consider that we have a LanceDB table named `my_table`, whose string column `text` we want to index and query via keyword search, the FTS index must be created before you can search via keywords.

=== "Python"

    ```python
    import lancedb

    uri = "data/sample-lancedb"
    db = lancedb.connect(uri)

    table = db.create_table(
        "my_table",
        data=[
            {"vector": [3.1, 4.1], "text": "Frodo was a happy puppy"},
            {"vector": [5.9, 26.5], "text": "There are several kittens playing"},
        ],
    )

>     # passing `use_tantivy=False` to use lance FTS index

Reviewer 2: do we have a default value for `use_tantivy`?

Lets show that in the code as well, i.e.,

```
# Create native FTS index
table.create_fts_index("text")
```

Author: Modified

    table.create_fts_index("text", use_tantivy=True)
    table.search("puppy").limit(10).select(["text"]).to_list()
    # [{'text': 'Frodo was a happy puppy', '_score': 0.6931471824645996}]
    # ...
    ```

=== "TypeScript"

    ```typescript

>     import * as lancedb from "@lancedb/lancedb";

Reviewer 2: do we need to call `createFtsIndex()` ?

Author: Added

    import lancedb

    uri = "data/sample-lancedb"
    db = lancedb.connect(uri)

    table = db.create_table(
        "my_table",
        data=[
            {"vector": [3.1, 4.1], "text": "Frodo was a happy puppy"},
            {"vector": [5.9, 26.5], "text": "There are several kittens playing"},
        ],
    )

    # passing `use_tantivy=False` to use lance FTS index
    table.create_fts_index("text", use_tantivy=True)
    table.search("puppy").limit(10).select(["text"]).to_list()
    # [{'text': 'Frodo was a happy puppy', '_score': 0.6931471824645996}]
    # ...
    ```

=== "TypeScript"

    ```typescript
    import * as lancedb from "@lancedb/lancedb";
    const uri = "data/sample-lancedb"
    const db = await lancedb.connect(uri);

    const data = [
    { vector: [3.1, 4.1], text: "Frodo was a happy puppy" },
    { vector: [5.9, 26.5], text: "There are several kittens playing" },
    ];
    const tbl = await db.createTable("my_table", data, { mode: "overwrite" });
    await tbl
    .search("puppy")
    .select(["text"])
    .limit(10)
    .toArray();
    ```

=== "Rust"

    ```rust
    let uri = "data/sample-lancedb";
    let db = connect(uri).execute().await?;

>     let initial_data: Box<dyn RecordBatchReader + Send> = create_some_records()?;
>     let tbl = db
>         .create_table("my_table", Box::new(initial_data))

Reviewer: `initial_data` is already a `Box`, don't need to re-wrap it.
```suggestion
    let initial_data: Box<dyn RecordBatchReader + Send> = create_some_records()?;
    let tbl = db
        .create_table("my_table", initial_data)
```

        .execute()

>         .await
>         .unwrap();

Reviewer: We use `?` already below, so might as well use it here.
```suggestion
        .await?;
```

    tbl
        .create_index(&["text"], Index::FTS(FtsIndexBuilder::default()))
        .execute()
        .await?;

    tbl
        .query()
        .full_text_search(FullTextSearchQuery::new("puppy".to_owned()))
        .select(lancedb::query::Select::Columns(vec!["text".to_owned()]))
        .limit(10)
        .execute()
        .await?;
    ```

It would search on all indexed columns by default, so it's useful when there are multiple indexed columns.
For now, this is supported in tantivy way only.

> Passing `fts_columns="text"` if you want to specify the columns to search, but it's not available for tantivy based full text search.

Reviewer: ```suggestion
Passing `fts_columns="text"` if you want to specify the columns to search, but it's not available for Tantivy-based full text search.
```

!!! note
    LanceDB automatically searches on the existing FTS index if the input to the search is of type `str`. If you provide a vector as input, LanceDB will search the ANN index instead.

## Tokenization
By default the text is tokenized by splitting on punctuation and whitespaces and then removing tokens that are longer than 40 chars. For more language specific tokenization then provide the argument tokenizer_name with the 2 letter language code followed by "_stem". So for english it would be "en_stem".

For now, only the tantivy based FTS index supports to specify the tokenizer, so it's only available in Python with `use_tantivy=True`.

=== "use_tantivy=True"

    ```python
    table.create_fts_index("text", use_tantivy=True, tokenizer_name="en_stem")
    ```

=== "use_tantivy=False"

>     ```python
>     not supported yet
>     ```

Reviewer: ```suggestion
    **Not supported yet**
```

the following [languages](https://docs.rs/tantivy/latest/tantivy/tokenizer/enum.Language.html) are currently supported.

## Index multiple columns

If you have multiple string columns to index, there's no need to combine them manually -- simply pass them all as a list to `create_fts_index`:

=== "use_tantivy=True"

    ```python
    table.create_fts_index(["text1", "text2"])
    ```

=== "use_tantivy=False"

>     ```python
>     not supported yet
>     ```

Reviewer: For these, could we link to a GH issue?
```suggestion
    **Not supported yet**
```

The FTS index must be created before you can search via keywords.

=== "> 0.11.0"

    ```python
    table.create_scalar_index("text", "INVERTED")
    ```

=== "<= 0.11.0"

    ```python
    table.create_fts_index("text")

>     ```

Reviewer 3: We could also combine the two create_index methods (scalar and vector).  The inverted index *can* be used in the same way as scalar indices (speeding up filters) but it can also be used for full text search (which is a search and not a filter)

>     ```

Author: I'm going to change the `create_fts_index(column: str)` to `create_fts_index(column: str, index_type="FTS", use_legacy=False)`. We haven't supported use the inverted index as filtering so it still makes sense.

the `create_index` seems has been used for vector index only, that can't be easily changed.

applied on top of the full text search results. This can be invoked via the familiar
`where` syntax:

=== "Python"

    ```python
    table.search("puppy").limit(10).where("meta='foo'").to_list()
    ```

=== "TypeScript"

    ```typescript
    await tbl
    .search("apple")
    .select(["id", "doc"])
    .limit(10)
    .where("meta='foo'")
    .toArray();
    ```

=== "Rust"

    ```rust
    table
        .query()
        .full_text_search(FullTextSearchQuery::new(words[0].to_owned()))
        .select(lancedb::query::Select::Columns(vec!["doc".to_owned()]))
        .limit(10)
        .only_if("meta='foo'")
        .execute()
        .await?;
    ```

## Sorting

> !!! tip "Warn"

Reviewer: It's confusing to use `tip` for a warning. Use `warning` instead:

```suggestion
!!! warning "Warn"
```

You can see all available admonitions here: https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types

>     Sorting is available for only tantivy based FTS

Reviewer: ```suggestion
    Sorting is available for only Tantivy-based FTS
```

## Phrase queries vs. terms queries

> !!! tip "Warn"

Reviewer: ```suggestion
!!! warning "Warn"
```
